cpp/src/io/orc/writer_impl.cu

```diff
@@ -1121,96 +1147,163 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
         rg_grp.num_chunks  = 1;
       }
     }
-    auto col_stats         = &file_stat_merge[column.index()];
-    col_stats->col         = stat_desc.device_ptr(column.index());
-    col_stats->start_chunk = static_cast<uint32_t>(column.index() * segmentation.num_stripes());
-    col_stats->num_chunks  = static_cast<uint32_t>(segmentation.num_stripes());
   }
   stat_desc.host_to_device(stream);
-  stat_merge.host_to_device(stream);
+  rowgroup_merge.host_to_device(stream);
+  stripe_merge.host_to_device(stream);
   set_stat_desc_leaf_cols(orc_table.d_columns, stat_desc, stream);
 
-  rmm::device_uvector<statistics_chunk> stat_chunks(num_stat_blobs, stream);
-  auto rowgroup_stat_chunks = stat_chunks.data();
-  auto stripe_stat_chunks   = rowgroup_stat_chunks + num_rowgroup_blobs;
-  auto file_stat_chunks     = stripe_stat_chunks + num_stripe_blobs;
+  // rowgroup stat chunks are written out in each stripe stripe and file-level chunks are written in
+  // the footer. to prevent persisting the rowgroup stat chunks across multiple calls to write in a
+  // chunked write situation, these allocations are split up so stripe data can persist until the
+  // footer write
+  rmm::device_uvector<statistics_chunk> rowgroup_chunks(num_rowgroup_blobs, stream);
+  rmm::device_uvector<statistics_chunk> stripe_chunks(num_stripe_blobs, stream);
+  auto rowgroup_stat_chunks = rowgroup_chunks.data();
+  auto stripe_stat_chunks   = stripe_chunks.data();
 
-  rmm::device_uvector<statistics_group> stat_groups(num_rowgroup_blobs, stream);
+  rmm::device_uvector<statistics_group> rowgroup_groups(num_rowgroup_blobs, stream);
   gpu::orc_init_statistics_groups(
-    stat_groups.data(), stat_desc.device_ptr(), segmentation.rowgroups, stream);
+    rowgroup_groups.data(), stat_desc.device_ptr(), segmentation.rowgroups, stream);
 
   detail::calculate_group_statistics<detail::io_file_format::ORC>(
-    stat_chunks.data(), stat_groups.data(), num_rowgroup_blobs, stream);
+    rowgroup_chunks.data(), rowgroup_groups.data(), num_rowgroup_blobs, stream);
 
   detail::merge_group_statistics<detail::io_file_format::ORC>(
-    stripe_stat_chunks,
-    rowgroup_stat_chunks,
-    stat_merge.device_ptr(num_rowgroup_blobs),
-    num_stripe_blobs,
-    stream);
+    stripe_stat_chunks, rowgroup_stat_chunks, stripe_merge.device_ptr(), num_stripe_blobs, stream);
 
-  detail::merge_group_statistics<detail::io_file_format::ORC>(
-    file_stat_chunks,
-    stripe_stat_chunks,
-    stat_merge.device_ptr(num_rowgroup_blobs + num_stripe_blobs),
-    num_file_blobs,
-    stream);
-  gpu::orc_init_statistics_buffersize(
-    stat_merge.device_ptr(), stat_chunks.data(), num_stat_blobs, stream);
-  stat_merge.device_to_host(stream, true);
+  // with chunked writes, the orc table can be deallocated between write calls.
+  // This forces our hand to encode row groups and stripes only in this stage and further
+  // we have to persist any data from the table that we need, min/max strings.
```

Comment:
I'm not sure if I understand this line. What does "min/max strings" mean at the end? Is that an example of the type of data that must be kept until the end of the chunked write? Are statistics always stored as strings?

Reply:
We are saving the strings stored in the statistics chunks, which are the minimum length string and the maximum length string. I'll clean this up as well, thanks for pointing it out.

Reply:
Reworded, please advise.